File: ProblemSolve/856.cpp
/************************
 * Score of Parentheses *
 ************************/

class Solution {
public:
    int scoreOfParentheses(string s) {
        stack<int> st;
        st.push(0);
        
        for(char ch : s){
            if(ch == '('){
                st.push(0);
            }
            else{
                int currValue = st.top();
                st.pop();
                
                int preValue = st.top();
                st.pop();
                
                st.push(preValue + max(2 * currValue, 1));
            }
                
        }
        
        return st.top();
    }
};

File: ProblemSolve/856.cs
/************************
 * Score of Parentheses *
 ************************/

public class Solution {
    public int ScoreOfParentheses(string s) {
        Stack<int> st = new Stack<int>();
        st.Push(0);
        
        foreach(char ch in s){
            if(ch == '('){
                st.Push(0);
            }
            else{
                int currVal = st.Pop();
                int preVal = st.Pop();
                
                st.Push(preVal + Math.Max(currVal * 2, 1));
            }
        }
        
        return st.Pop();
    }
}